File: MCP/mcp_notes/NotesMcpServer.java
/**
 * Notes MCP Server — Java implementation
 * ========================================
 * A minimal but complete MCP server that demonstrates:
 *   - Tool registration  (add_note, list_notes, get_note)
 *   - Resource exposure  (notes://all)
 *   - Prompt templates   (summarize_notes)
 *   - stdio transport    (default for local servers)
 *
 * Dependencies (Maven):
 * ─────────────────────
 * <dependency>
 *     <groupId>io.modelcontextprotocol.sdk</groupId>
 *     <artifactId>mcp</artifactId>
 *     <version>0.9.0</version>   <!-- use latest from Maven Central -->
 * </dependency>
 *
 * Build & run:
 *   mvn compile exec:java -Dexec.mainClass="NotesMcpServer"
 *
 * Or with the MCP Inspector (requires Node):
 *   npx @modelcontextprotocol/inspector java -cp target/classes NotesMcpServer
 */

import io.modelcontextprotocol.server.McpServer;
import io.modelcontextprotocol.server.McpSyncServer;
import io.modelcontextprotocol.server.transport.StdioServerTransportProvider;
import io.modelcontextprotocol.spec.McpSchema;

import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class NotesMcpServer {

    // ─────────────────────────────────────────────
    // 1. In-memory data model
    // ─────────────────────────────────────────────

    record Note(int id, String title, String content, String createdAt) {}

    private static final List<Note> notes = new ArrayList<>();
    private static final AtomicInteger idCounter = new AtomicInteger(0);

    private static String now() {
        return DateTimeFormatter.ISO_INSTANT.format(Instant.now());
    }

    // ─────────────────────────────────────────────
    // 2. Helper: build the full notes Markdown
    // ─────────────────────────────────────────────

    private static String buildAllNotesMarkdown() {
        if (notes.isEmpty()) return "_No notes saved yet._";
        return notes.stream()
                .map(n -> "## " + n.title() + "\n"
                        + "_id=" + n.id() + " | saved: " + n.createdAt() + "_\n\n"
                        + n.content())
                .collect(Collectors.joining("\n\n---\n\n"));
    }

    // ─────────────────────────────────────────────
    // 3. Tool handlers
    // ─────────────────────────────────────────────

    /** add_note: saves a new note, returns confirmation. */
    private static McpSchema.CallToolResult handleAddNote(Map<String, Object> args) {
        String title   = (String) args.get("title");
        String content = (String) args.get("content");

        if (title == null || content == null) {
            return errorResult("Both 'title' and 'content' are required.");
        }

        Note note = new Note(idCounter.incrementAndGet(), title, content, now());
        notes.add(note);

        return McpSchema.CallToolResult.builder()
                .addTextContent("✓ Note saved (id=" + note.id() + "): \"" + title + "\"")
                .build();
    }

    /** list_notes: returns a summary table of all notes. */
    private static McpSchema.CallToolResult handleListNotes() {
        if (notes.isEmpty()) {
            return McpSchema.CallToolResult.builder()
                    .addTextContent("No notes yet.")
                    .build();
        }
        String summary = notes.stream()
                .map(n -> "id=" + n.id() + " | " + n.createdAt() + " | " + n.title())
                .collect(Collectors.joining("\n"));
        return McpSchema.CallToolResult.builder()
                .addTextContent(summary)
                .build();
    }

    /** get_note: returns full content of a note by id. */
    private static McpSchema.CallToolResult handleGetNote(Map<String, Object> args) {
        Object idArg = args.get("id");
        if (idArg == null) return errorResult("'id' is required.");

        int id = ((Number) idArg).intValue();
        return notes.stream()
                .filter(n -> n.id() == id)
                .findFirst()
                .map(n -> McpSchema.CallToolResult.builder()
                        .addTextContent("# " + n.title() + "\n_Saved: "
                                + n.createdAt() + "_\n\n" + n.content())
                        .build())
                .orElse(errorResult("Note id=" + id + " not found."));
    }

    private static McpSchema.CallToolResult errorResult(String message) {
        return McpSchema.CallToolResult.builder()
                .addTextContent("Error: " + message)
                .isError(true)
                .build();
    }

    // ─────────────────────────────────────────────
    // 4. main — wire up and start the server
    // ─────────────────────────────────────────────

    public static void main(String[] args) {

        // 4a. Transport: stdio (stdin → server, server → stdout)
        var transportProvider = new StdioServerTransportProvider();

        // 4b. Server info advertised during initialization handshake
        var serverInfo = new McpSchema.Implementation("notes-server", "1.0.0");

        // 4c. Build the synchronous MCP server
        //     (McpSyncServer wraps async internals with a simpler blocking API)
        McpSyncServer server = McpServer.sync(transportProvider)
                .serverInfo(serverInfo)

                // ── TOOLS ───────────────────────────────────────────────────

                .tool(
                    McpSchema.Tool.builder()
                        .name("add_note")
                        .description("Save a new note with a title and content. " +
                                     "Use whenever the user wants to remember something.")
                        .inputSchema(McpSchema.JsonSchema.builder()
                            .type("object")
                            .properties(Map.of(
                                "title",   Map.of("type", "string",
                                                  "description", "Short descriptive title."),
                                "content", Map.of("type", "string",
                                                  "description", "Full body of the note.")
                            ))
                            .required(List.of("title", "content"))
                            .build())
                        .build(),
                    (exchange, toolArgs) -> handleAddNote(toolArgs.arguments())
                )

                .tool(
                    McpSchema.Tool.builder()
                        .name("list_notes")
                        .description("Return a summary of all saved notes (id, title, timestamp).")
                        .inputSchema(McpSchema.JsonSchema.builder()
                            .type("object")
                            .properties(Map.of())
                            .required(List.of())
                            .build())
                        .build(),
                    (exchange, toolArgs) -> handleListNotes()
                )

                .tool(
                    McpSchema.Tool.builder()
                        .name("get_note")
                        .description("Retrieve the full content of a note by its id.")
                        .inputSchema(McpSchema.JsonSchema.builder()
                            .type("object")
                            .properties(Map.of(
                                "id", Map.of("type", "integer",
                                             "description", "Numeric id of the note.")
                            ))
                            .required(List.of("id"))
                            .build())
                        .build(),
                    (exchange, toolArgs) -> handleGetNote(toolArgs.arguments())
                )

                // ── RESOURCES ───────────────────────────────────────────────

                .resource(
                    McpSchema.Resource.builder()
                        .uri("notes://all")
                        .name("All notes")
                        .description("Every saved note rendered as a Markdown document.")
                        .mimeType("text/markdown")
                        .build(),
                    (exchange, request) -> {
                        String markdown = buildAllNotesMarkdown();
                        return McpSchema.ReadResourceResult.builder()
                                .addTextContent(request.uri(), "text/markdown", markdown)
                                .build();
                    }
                )

                // ── PROMPT TEMPLATES ─────────────────────────────────────────

                .prompt(
                    McpSchema.Prompt.builder()
                        .name("summarize_notes")
                        .description("Produce a concise summary of all saved notes.")
                        .arguments(List.of())
                        .build(),
                    (exchange, request) -> {
                        String allNotes = buildAllNotesMarkdown();
                        return McpSchema.GetPromptResult.builder()
                                .description("Summarize all notes")
                                .addUserMessage(
                                    "Here are my notes:\n\n" + allNotes +
                                    "\n\nPlease give me a concise bullet-point summary."
                                )
                                .build();
                    }
                )

                .build();

        // 4d. Keep the server running until stdin closes
        //     (the MCP client controls the lifecycle)
        System.err.println("Notes MCP Server running on stdio. Waiting for client...");
        server.closeGracefully();
    }
}

File: MCP/mcp_notes/notes_mcp_server.py
"""
Notes MCP Server — Python implementation
=========================================
A minimal but complete MCP server that demonstrates:
  - Tool registration (add_note, list_notes)
  - Resource exposure (notes://all)
  - Prompt templates (summarize_notes)
  - stdio transport (default for local servers)

Requirements:
    pip install mcp

Run directly (stdio transport — connect from Claude Desktop or mcp CLI):
    python notes_mcp_server.py

Test with the MCP CLI:
    pip install mcp[cli]
    mcp dev notes_mcp_server.py
"""

from datetime import datetime
from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp import types
import asyncio

# ─────────────────────────────────────────────
# 1. In-memory data store
# ─────────────────────────────────────────────

notes: list[dict] = []  # Each note: { id, title, content, created_at }


def _next_id() -> int:
    return len(notes) + 1


# ─────────────────────────────────────────────
# 2. Create the MCP Server instance
# ─────────────────────────────────────────────

app = Server("notes-server")


# ─────────────────────────────────────────────
# 3. Tool definitions
#    The LLM sees the name + description + inputSchema
#    and decides when to call them.
# ─────────────────────────────────────────────

@app.list_tools()
async def list_tools() -> list[types.Tool]:
    """Advertise available tools to the MCP client."""
    return [
        types.Tool(
            name="add_note",
            description=(
                "Save a new note with a title and content. "
                "Use this whenever the user wants to remember something."
            ),
            inputSchema={
                "type": "object",
                "properties": {
                    "title": {
                        "type": "string",
                        "description": "A short, descriptive title for the note.",
                    },
                    "content": {
                        "type": "string",
                        "description": "The full body/content of the note.",
                    },
                },
                "required": ["title", "content"],
            },
        ),
        types.Tool(
            name="list_notes",
            description=(
                "Return a summary list of all saved notes (id, title, timestamp). "
                "Use this to show the user what notes exist."
            ),
            inputSchema={
                "type": "object",
                "properties": {},
                "required": [],
            },
        ),
        types.Tool(
            name="get_note",
            description=(
                "Retrieve the full content of a specific note by its id."
            ),
            inputSchema={
                "type": "object",
                "properties": {
                    "id": {
                        "type": "integer",
                        "description": "The numeric id of the note to retrieve.",
                    }
                },
                "required": ["id"],
            },
        ),
    ]


@app.call_tool()
async def call_tool(
    name: str, arguments: dict
) -> list[types.TextContent | types.ImageContent | types.EmbeddedResource]:
    """
    Dispatch incoming tool calls to handler functions.
    Must return a list of content blocks.
    """
    if name == "add_note":
        title = arguments["title"]
        content = arguments["content"]
        note = {
            "id": _next_id(),
            "title": title,
            "content": content,
            "created_at": datetime.now().isoformat(timespec="seconds"),
        }
        notes.append(note)
        return [
            types.TextContent(
                type="text",
                text=f"✓ Note saved (id={note['id']}): "{title}"",
            )
        ]

    elif name == "list_notes":
        if not notes:
            return [types.TextContent(type="text", text="No notes yet.")]
        lines = [f"id={n['id']} | {n['created_at']} | {n['title']}" for n in notes]
        return [types.TextContent(type="text", text="\n".join(lines))]

    elif name == "get_note":
        note_id = arguments["id"]
        note = next((n for n in notes if n["id"] == note_id), None)
        if note is None:
            return [types.TextContent(type="text", text=f"Note id={note_id} not found.")]
        text = f"# {note['title']}\n_Saved: {note['created_at']}_\n\n{note['content']}"
        return [types.TextContent(type="text", text=text)]

    else:
        raise ValueError(f"Unknown tool: {name}")


# ─────────────────────────────────────────────
# 4. Resource definitions
#    Resources are data sources the client can
#    read and inject into the model's context.
# ─────────────────────────────────────────────

@app.list_resources()
async def list_resources() -> list[types.Resource]:
    """Advertise available resources."""
    return [
        types.Resource(
            uri="notes://all",
            name="All notes",
            description="Every saved note as a Markdown document.",
            mimeType="text/markdown",
        )
    ]


@app.read_resource()
async def read_resource(uri: str) -> str:
    """Return the content of a requested resource URI."""
    if uri == "notes://all":
        if not notes:
            return "_No notes saved yet._"
        sections = []
        for n in notes:
            sections.append(
                f"## {n['title']}\n"
                f"_id={n['id']} | saved: {n['created_at']}_\n\n"
                f"{n['content']}"
            )
        return "\n\n---\n\n".join(sections)
    raise ValueError(f"Unknown resource URI: {uri}")


# ─────────────────────────────────────────────
# 5. Prompt templates
#    Pre-built prompt patterns the client can
#    fetch and inject into the conversation.
# ─────────────────────────────────────────────

@app.list_prompts()
async def list_prompts() -> list[types.Prompt]:
    """Advertise available prompt templates."""
    return [
        types.Prompt(
            name="summarize_notes",
            description="Produce a concise summary of all saved notes.",
            arguments=[],
        )
    ]


@app.get_prompt()
async def get_prompt(
    name: str, arguments: dict | None
) -> types.GetPromptResult:
    """Return a rendered prompt template."""
    if name == "summarize_notes":
        all_notes_md = await read_resource("notes://all")
        return types.GetPromptResult(
            description="Summarize all notes",
            messages=[
                types.PromptMessage(
                    role="user",
                    content=types.TextContent(
                        type="text",
                        text=(
                            "Here are my notes:\n\n"
                            f"{all_notes_md}\n\n"
                            "Please give me a concise bullet-point summary."
                        ),
                    ),
                )
            ],
        )
    raise ValueError(f"Unknown prompt: {name}")


# ─────────────────────────────────────────────
# 6. Entry point — run over stdio transport
# ─────────────────────────────────────────────

async def main():
    async with stdio_server() as (read_stream, write_stream):
        await app.run(
            read_stream,
            write_stream,
            app.create_initialization_options(),
        )


if __name__ == "__main__":
    asyncio.run(main())
